### DISCUSSION TITLE
Can I have an authentication adapter without a backend?

### DISCUSSION THREAD

**Discussion by adrianotadao** [rainbow-me/rainbowkit#885] [Q&A]
Hey folks, it may be a stupid question but I would love your help to understand something.

**Context**
My application needs a wallet connect button and I would like to use rainbowkit for it but I also need to force the user to sign a message before authenticating them. For that, I think the only way of doing it is using custom authentication (https://www.rainbowkit.com/docs/custom-authentication). That said, I'm wondering why the backend is important here since this is not that clear to me and it seems that all of it could be done in the frontend only as follows:

```javascript
const authenticationAdapter = useMemo(
  () =>
    createAuthenticationAdapter({
      getNonce: async () => Math.floor(Math.random() * 10000 * 10000).toString(),

      createMessage: ({ address, chainId }) => {
        const message = new SiweMessage({
          domain: window.location.host,
          address,
          statement: 'Potato Login.',
          uri: window.location.origin,
          version: '1',
          chainId,
        });

        return message;
      },

      getMessageBody: ({ message }) => message.toMessage(),

      verify: async ({ message, signature }) => {
        if (!address) return Boolean(false);

        const siweMessage = new SiweMessage(message.toMessage());

        try {
          const response = await siweMessage.verify({ signature });

          if (address !== response.data.address) return false;

          return true;
        } catch {
          return false;
        }
      },

      signOut: async () => {},
    }),
  [address],
);

return (
  <RainbowKitAuthenticationProvider adapter={authenticationAdapter} status={loginStatus}>
    <RainbowKitProvider>
      ...
    </RainbowKitProvider>
  </RainbowKitAuthenticationProvider>
);
```

**Questions**

**1** - What would be the problem if I don't have a backend here? Would the backend only be useful to avoid replay attacks like [here](https://wagmi.sh/examples/sign-in-with-ethereum) where it checks if the nonce in the session is the same in the message?

**2** - Could a user, in my code example, somehow impersonate another user if they have the public address or something?

Thank you very much, folks!

**Comment by aalmada** [comment]
The authentication provider is used to create a session on a backend. This is based on https://eips.ethereum.org/EIPS/eip-4361. It assumes there's a backend.

If you only want to sign a message, you should check wagmi: https://wagmi.sh/examples/sign-message. RainbowKit is based on wagmi so implicitly your app already depends on it.

**Reply by adrianotadao** [reply]
thx @aalmada for answering this. Could I ask you to take a look at my question number **2**, please? I'm a little bit concerned if that can be dangerous or not. Thank you very much!

**Reply by aalmada** [reply]
@adrianotadao I rewrote the answer for this one https://github.com/rainbow-me/rainbowkit/discussions/797#discussioncomment-4267206
Does this help?
